### PR TITLE
Closes #1542: Can't go back out of YouTube

### LINKED ISSUE #1542: Can't go back out of YouTube (with remote back)
## Steps to reproduce
1. Click Google search tile
2. Open overlay and click YouTube tile
3. Click back button on remote several times to try and get back to Google
### Expected behavior
Can get back to Google
### Actual behavior
Cannot go back to Google from the remote back button. Focus stays on the 'Home' icon of YouTube.
### Device information

* Fire TV device: 4k
* Firefox version: 3.1

### PR DESCRIPTION
**Pull request by dnarcese**
### Pull Request checklist

- [x] This PR includes thorough **tests** or an explanation of why it does not
In the interest of time, this will be covered in #1420.
- [x] This PR includes a **CHANGELOG entry** or does not need one

### CODE REVIEW

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 1]
```diff
@@ -284,6 +289,8 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
         if (session.url == APP_URL_HOME) {
             browserOverlay?.visibility = View.VISIBLE
         }
+
+        webView.addJavascriptInterface(YouTubeBackHandler(), "youtubeBackHandler")
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: `youtubeBackHandler` gets written directly into the DOM so you probably want to name it something:
- identifiable by website owners wondering what this random object is
- identifiable by us to know we added it
- that won't get clobbered by site content

[For media session, we use `_firefoxTV_playbackStateObserverJava`](https://github.com/mozilla-mobile/firefox-tv/blob/23a1288990e22a3fc06a2975bac0c307775a9e8d/app/src/main/java/org/mozilla/tv/firefox/webrender/VideoVoiceCommandMediaSession.kt#L52).

By the way, see `VideoVoiceCommandMediaSession` for an example of how you can break a JavascriptInterface dependent class into a separate file.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 2]
```diff
@@ -421,9 +428,10 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
         }
 
         if (!browserOverlay.isVisible && session.isYoutubeTV &&
-                event.keyCode == KeyEvent.KEYCODE_BACK) {
-            val escKeyEvent = KeyEvent(event.action, KeyEvent.KEYCODE_ESCAPE)
-            (activity as MainActivity).dispatchKeyEvent(escKeyEvent)
+                event.keyCode == KeyEvent.KEYCODE_BACK &&
+                !youtubeOverride) {
+            youtubeKeyEvents.add(event)
+            webView?.evalJS("youtubeBackHandler.youtubeBack(document.activeElement.parentElement.parentElement.id === 'guide-list');")
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Did you verify this is the best way to identify if we're at the top level in YouTube's back stack? I did this rather quickly.

Also, we should clarify what `document.active.parent.parent` actually means and why it works, preferably in code but in comments if that fails.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 3]
```diff
@@ -447,6 +455,44 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+    /**
+     * JavascriptInterface for handling when to map back click to ESC vs perform back in history.
+     * Clicking remote back will remap to ESC until the active DOM element is an item in YouTube's guidelist-
+     * then it will go back in the history stack.
+     */
+    inner class YouTubeBackHandler {
+        @JavascriptInterface
+        fun youtubeBack(sendBack: Boolean) {
+            if (sendBack) {
+                Handler(Looper.getMainLooper()).post {
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Creating a new Handler might be expensive so we wouldn't want to do it each time this is called. Maybe `launch(UI)`?

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 4]
```diff
@@ -447,6 +455,44 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                    val wv = (webView as ViewGroup).getChildAt(0) as WebView
+                    val backForward = wv.copyBackForwardList().toList()
+                    val youtubeIndex = backForward.lastIndexOf("https://ftv.cdn.mozilla.net/ytht")
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: This isn't always correct: we need to find this or `youtube.com/tv`, e.g. if the user navigates there directly.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 5]
```diff
@@ -447,6 +455,44 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                    wv.goBackOrForward(-goBackSteps)
+                }
+                return
+            }
+            val keyCode = if (sendBack) KeyEvent.KEYCODE_BACK else KeyEvent.KEYCODE_ESCAPE
+
+            youtubeOverride = true
+            youtubeKeyEvents.forEach {
+                val newKeyEvent = KeyEvent(it.action, keyCode)
+                Handler(Looper.getMainLooper()).post {
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Again, new handler created.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 6]
```diff
@@ -447,6 +455,44 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                    val backForward = wv.copyBackForwardList().toList()
+                    val youtubeIndex = backForward.lastIndexOf("https://ftv.cdn.mozilla.net/ytht")
+                    val goBackSteps = backForward.size - youtubeIndex
+                    wv.goBackOrForward(-goBackSteps)
+                }
+                return
+            }
+            val keyCode = if (sendBack) KeyEvent.KEYCODE_BACK else KeyEvent.KEYCODE_ESCAPE
+
+            youtubeOverride = true
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: I wonder if we can do this without the youtubeOverride? I did it because simulating the `ESC` presses called this method again so I needed a flag to prevent nested calls but I wonder if there is a better way.

File: CHANGELOG.md [hunk 7]
```diff
@@ -10,7 +10,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Changed
 
 ### Fixed
-- Some web fonts were blocked which caused icons to be missing
```
Comment (mcomella) [inline CHANGELOG.md]: I don't think we should delete this line, right?

File: CHANGELOG.md [hunk 8]
```diff
@@ -10,7 +10,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
+- Unable to back out of YouTube with remote back button
```
Comment (mcomella) [inline CHANGELOG.md]: "An issue where users were unable to leave YouTube with remote back button (#...)"

Include a bug number and perhaps elaborate more, like ^

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 9]
```diff
@@ -421,9 +428,10 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
         }
 
         if (!browserOverlay.isVisible && session.isYoutubeTV &&
-                event.keyCode == KeyEvent.KEYCODE_BACK) {
-            val escKeyEvent = KeyEvent(event.action, KeyEvent.KEYCODE_ESCAPE)
-            (activity as MainActivity).dispatchKeyEvent(escKeyEvent)
+                event.keyCode == KeyEvent.KEYCODE_BACK &&
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Did you try this with the browser navigation overlay back button too?

**Comment by mcomella** [comment]
@dnarcese Looking at my comments, I think it seems reasonable to _attempt_ to address everything except trying to abstract the code away to a one file, which can take a while given the uncertainty of design. Since the code works, if something seems like it'll take too long and we'll miss our deadline, **feel free to skip it** and make a note!

To summarize the issues to address, in order of priority:
- Pressing back from Pocket can exit YouTube (not sure if this already works; note that it does not need to go back to Pocket, just needs to exit YouTube)
- Try to have a better system than boolean mutable flags and key event lists
- Ensure we're using an optimal way to ensure we're at the top level in YouTube's ESC back stack
- Don't create new handlers on each button press (cache a handler or use coroutines)
- Disambiguate the object injected into the DOM with a better name
- Address changelog nit

We should ask Ashley if these use cases are important and if so, we should fix them too (probably at the highest priority):
- Overlay back button can also leave YouTube
- We can also back out if the user visits youtube.com/tv by typing the URL (this might be the same as the Pocket use case)

Let me know if you have questions!

**Comment by dnarcese** [comment]
@mcomella I've addressed everything from above except for 

- Pressing back from Pocket can exit YouTube
- We can also back out if the user visits youtube.com/tv by typing the URL

Those 2 issues are trickier to deal with and it was decided that since more users click the YouTube home tile than the Pocket megatile, this issue is lower priority and can be skipped for now.

**Comment by liuche** [comment]
@dnarcese can you file a bug for the follow-ups, and document the problems with the Pocket youtube redirect back stack, and optionally ideas for how to address them?

File: CHANGELOG.md [hunk 10]
```diff
@@ -11,6 +11,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 ### Changed
 
 ### Fixed
+- Issue where users were unable to exit YouTube with remote back button (#1542)
```
Comment (mcomella) [inline CHANGELOG.md]: Not your fault: the release date for 3.1.3 is inaccurate and the Changed header should only be present if there are items under it

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 11]
```diff
@@ -409,9 +413,26 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                event.action == KeyEvent.ACTION_DOWN) {
+            webView?.evalJS("""
+                (function () {
+                    return document.activeElement.parentElement.parentElement.id === 'guide-list';
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: nit: comment to explain what is happening here. Probably outside of the JS because it doesn't make sense to inject JS comments.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 12]
```diff
@@ -409,9 +413,26 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                })();
+                """,
+                    ValueCallback {
+                        val keyCode = if (it == "true") {
+                            val wv = (webView as ViewGroup).getChildAt(0) as WebView
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: nit: This is a lot of nesting: can we define the `ValueCallback` elsewhere so that it isn't as heavily nested?

Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: nit: We avoid one or two letter variable names. -> `webView`?

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 13]
```diff
@@ -409,9 +413,26 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                            val backForward = wv.copyBackForwardList().toStringList()
+                            val youtubeIndex = backForward.lastIndexOf("https://ftv.cdn.mozilla.net/ytht")
+                            val goBackSteps = backForward.size - youtubeIndex
+                            wv.goBackOrForward(-goBackSteps)
+                            KeyEvent.KEYCODE_BACK
+                        } else KeyEvent.KEYCODE_ESCAPE
+
+                        val newKeyEvent = KeyEvent(KeyEvent.ACTION_UP, keyCode)
+                        (activity as MainActivity).dispatchKeyEvent(newKeyEvent)
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: iiuc, I see a few problems with this:
- We're not dispatching the down event. That might work in this case but I'm concerned what side effects there might be
- We're dispatching the up event twice.
- We're not converting the BACK, UP physical key press into an ESC, UP physical key press

fwiw, I know we're pressed for time but I think it's think it's these types of edge cases that get uncovered while writing tests for code like this.

To be explicit, I think the logic we want here is:
```
if (youtube.com/tv) {
    if (!atYouTubeTopLevel) {
        dispatchHomeEvent(event.action)
    } else {
        goBackBeforeYouTube()
    }
} else {
    dispatchKeyEventAsNormal()
}
```

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 14]
```diff
@@ -409,9 +413,26 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+            webView?.evalJS("""
+                (function () {
+                    return document.activeElement.parentElement.parentElement.id === 'guide-list';
+                })();
+                """,
+                    ValueCallback {
+                        val keyCode = if (it == "true") {
+                            val wv = (webView as ViewGroup).getChildAt(0) as WebView
+                            val backForward = wv.copyBackForwardList().toStringList()
+                            val youtubeIndex = backForward.lastIndexOf("https://ftv.cdn.mozilla.net/ytht")
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: nit: Is this url stored in a constant somewhere? If not, can we add it as one?

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 15]
```diff
@@ -409,9 +413,26 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                event.action == KeyEvent.ACTION_DOWN) {
+            webView?.evalJS("""
+                (function () {
+                    return document.activeElement.parentElement.parentElement.id === 'guide-list';
+                })();
+                """,
+                    ValueCallback {
+                        val keyCode = if (it == "true") {
+                            val wv = (webView as ViewGroup).getChildAt(0) as WebView
+                            val backForward = wv.copyBackForwardList().toStringList()
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: nit: `backForwardUrls` or similar

File: app/src/main/java/org/mozilla/tv/firefox/ext/WebBackForwardList.kt [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.ext
+
+import android.webkit.WebBackForwardList
+
+/**
+ * Convert WebBackForwardList to List<String> so it is easier to use.
+ */
+fun WebBackForwardList.toStringList(): List<String> {
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/ext/WebBackForwardList.kt]: nit: `toUrlStringList` or similar: we want to be explicit that we're extracting the URLs.

However, I think I'd prefer to leave it as, `WebBackForwardList.toList()` and the callers can extract the string if they want: it's more flexible this way.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 17]
```diff
@@ -409,9 +413,26 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                            val youtubeIndex = backForward.lastIndexOf("https://ftv.cdn.mozilla.net/ytht")
+                            val goBackSteps = backForward.size - youtubeIndex
+                            wv.goBackOrForward(-goBackSteps)
+                            KeyEvent.KEYCODE_BACK
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Do we need to dispatch this back press at all if we're already going back before YouTube? I don't think so but I could be mistaken.

**Comment by mcomella** [comment]
If we're pressed for time, everything but the logic correction can slip.

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 18]
```diff
@@ -411,12 +416,40 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
         if (!browserOverlay.isVisible && session.isYoutubeTV &&
                 event.keyCode == KeyEvent.KEYCODE_BACK) {
             val escKeyEvent = KeyEvent(event.action, KeyEvent.KEYCODE_ESCAPE)
-            (activity as MainActivity).dispatchKeyEvent(escKeyEvent)
+
+            /**
+             * If YouTube guide-list element is focused:
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: nit: the comments in here don't explain what problem we're trying to solve and, because our solution is complex, the solution at a high level. If I came across this code for the first time, I might be confused at why this BACK->ESC override is necessary (i.e. YouTube doesn't handle their back stack correctly) and why the override is only necessary _sometimes_.

For example, I might comment, "youtube/tv doesn't handle their back stack correctly: a user can navigate back by pressing ESC but going back in the browser back stack does nothing. To fix this, we convert BACK presses on youtube/tv to ESC. However, when we get to the top of YouTube's back stack, we back all the way out of YouTube."

File: app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt [hunk 19]
```diff
@@ -411,12 +416,40 @@ class WebRenderFragment : EngineViewLifecycleFragment(), Session.Observer {
...
+                    jsCallback)
             return true
         }
         return false
     }
 
+    private fun goBackBeforeYouTube() {
+        val webView = (webView as ViewGroup).getChildAt(0) as WebView
+        val backForwardUrlList = webView.copyBackForwardList().toList().map { it.originalUrl }
+        val youtubeIndex = backForwardUrlList.lastIndexOf(YOUTUBE_TILE_URL)
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/webrender/WebRenderFragment.kt]: Finding the index and going back the appropriate number of steps would be a good candidate for writing tests, e.g.:
- How does this handle youtube.com/tv URLs?
- What happens if the CDN url is not in the format we expect (no trailing slash)?
- What happens if the CDN is at the end of the url list?
- What happens if the CDN url is not in the list?
- What if the CDN url is first?
- What if we remove firefox:home from the pages loaded by the application?

You get the idea. :P
